sklearn/cluster/_hdbscan/_tree.pyx

```diff
@@ -448,46 +413,42 @@ cdef cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] do_labelling(
     cdef:
         cnp.intp_t root_cluster
         cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] result
-        cnp.intp_t[:] parent_array, child_array
-        cnp.float64_t[:] lambda_array
+        cnp.ndarray[cnp.intp_t, ndim=1] parent_array, child_array
+        cnp.ndarray[cnp.float64_t, ndim=1] lambda_array
         TreeUnionFind union_find
-        cnp.intp_t n, parent, child, cluster
+        cnp.intp_t n, parent, child, cluster, label
+        cnp.float64_t parent_lambda, max_child_lambda
```

Comment:
Can we revert changes to `do_labeling`, because this is being worked on in https://github.com/scikit-learn/scikit-learn/pull/26101?

This way, this PR can focus on removing the need for sorting.